src/main/drivers/compass/compass_ist8310.c

```diff
@@ -0,0 +1,203 @@
+ *         0   |   0  |  0   |  1   |   Single measurement mode
+ *                                       Others: Reserved
+ *
+ * CTRL_REGB: Control Register 2
+ * Read Write
+ * Default value: 0x0B
+ * 7:4  0   Reserved.
+ * 3    DREN : Data ready enable control:
+ *      0: disable
+ *      1: enable
+ * 2    DRP: DRDY pin polarity control
+ *      0: active low
+ *      1: active high
+ * 1    0   Reserved.
+ * 0    SRST: Soft reset, perform Power On Reset (POR) routine
+ *      0: no action
+ *      1: start immediately POR routine
+ *      This bit will be set to zero after POR routine
+ */
+
+#define IST8310_REG_DATA 0x03
+#define IST8310_REG_WAI 0x00
+#define IST8310_REG_WAI_VALID 0x10
+
+#define IST8310_REG_STAT1 0x02
+#define IST8310_REG_STAT2 0x09
+
+#define IST8310_DRDY_MASK 0x01
+
+// I2C Control Register
+#define IST8310_REG_CNTRL1 0x0A
+#define IST8310_REG_CNTRL2 0x0B
+#define IST8310_REG_AVERAGE 0x41
+#define IST8310_REG_PDCNTL 0x42
+
+// Parameter
+// ODR = Output Data Rate, we use single measure mode for getting more data.
+#define IST8310_ODR_SINGLE 0x01
+#define IST8310_ODR_10_HZ 0x03
+#define IST8310_ODR_20_HZ 0x05
+#define IST8310_ODR_50_HZ 0x07
+#define IST8310_ODR_100_HZ 0x06
+
+#define IST8310_AVG_16  0x24
+#define IST8310_PULSE_DURATION_NORMAL 0xC0
+
+#define IST8310_CNTRL2_RESET 0x01
+#define IST8310_CNTRL2_DRPOL 0x04
+#define IST8310_CNTRL2_DRENA 0x08
+
+static bool ist8310Init(magDev_t *magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+    uint8_t regTemp;
+
+    busDeviceRegister(dev);
+
+    // soft-Reset
+    bool ack = busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp |= IST8310_CNTRL2_RESET;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(30);
+
+    // Init setting
+    ack = ack && busWriteRegister(dev, IST8310_REG_AVERAGE, IST8310_AVG_16);
+    delay(6);
+
+    busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_SINGLE);
+
+    return ack;
+}
+
+static bool ist8310Read(magDev_t * magDev, int16_t *magData)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    uint8_t buf[6];
+    uint8_t LSB2FSV = 3; // 3mG - 14 bit
+
+    static enum {
+        STATE_REQUEST_DATA,
+        STATE_FETCH_DATA,
+    } state = STATE_REQUEST_DATA;
+
+    switch (state) {
+        default:
+        case STATE_REQUEST_DATA:
+            busReadRegisterBuffer(dev, IST8310_REG_DATA, buf, sizeof(buf));
+
+            state = STATE_FETCH_DATA;
+
+            return false;
+        case STATE_FETCH_DATA:
+            // Looks like datasheet is incorrect and we need to invert Y axis to conform to right hand rule
+            magData[X] =  (int16_t)(buf[1] << 8 | buf[0]) * LSB2FSV;
+            magData[Y] = -(int16_t)(buf[3] << 8 | buf[2]) * LSB2FSV;
+            magData[Z] =  (int16_t)(buf[5] << 8 | buf[4]) * LSB2FSV;
+
+            busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_SINGLE);
+
+            state = STATE_REQUEST_DATA;
+
+            return true;
+    }
```

Comment:
This does not work because the buf is not static. The buf is read in one state and is processed in a different one, but the value has been lost because is not static.
For other place, the calculation of the magData has a time negligible, so maybe is better to move it to the REQUEST_DATA state, and have two states, more simple:
- The first one asks for the ODR single.
- The seconde one reads and processes it.

I will do more tests with the `busReadRegisterBufferStart` to see if I can get it to work. If it works, then better let the code as you have now.